### PR TITLE
[WIP] Making pbs_asynrunjob truely asynchronous

### PR DESCRIPTION
**Pull request by agrawalravi90**
#### Describe Bug or Feature

Scheduler today seems to spend the majority of its time waiting for an ACK from server for its runjob request. For my test setup with system with 50k jobs which can all run, the sched cycle took 172 seconds, out of which 161 seconds, or 94% of total time was spent by PBSD_rdrpy(), which is what waits for an ACK from server. pbs_asyrunjob() doesn't really do anything meaningful with the reply that it gets from the server, it frees it in the very next line. So, the purpose of PBSD_rdrpy() call seems to be to just make sure that the server got it, but since we use TCP to communicate, this shouldn't be necessary. We anyways don't send the runjob request again if we don't get a reply from server, so it really doesn't provide much value today, and removing it would make the scheduler much faster.

#### Describe Your Change

Removed the call to PBSD_rdrpy() from pbs_asyrunjob(), and added checks in server so that it won't send a reply for pbs_asyrunjob() requests.

**Performance numbers:**
50k jobs, 50k ncpus, time taken by 1 sched cycle:
before: 3 minutes
after: 12 seconds 

#### Attach Test and Valgrind Logs/Output

This is refactoring, so running existing tests should suffice.

### CODE REVIEW

**Comment by arungrover** [comment]
Hey Ravi,
I think if we do this then the server will stop sending a reply back in the case when a runjob hook rejects a job. While the scheduler does nothing with the reply, it does need the error number in order to identify that a runjob hook has rejected (or there is a protocol failure) the request and scheduler won't account for the job's requested resources.

**Comment by agrawalravi90** [comment]
> Hey Ravi,
I think if we do this then the server will stop sending a reply back in the case when a runjob hook rejects a job. While the scheduler does nothing with the reply, it does need the error number in order to identify that a runjob hook has rejected (or there is a protocol failure) the request and scheduler won't account for the job's requested resources.

Thanks for your reply Arun. If I understand your concern correctly, then this is similar to the mom rejecting a job, which will cause the same issue today, so is that a big issue? The next sched cycle, which will hopefully be sooner after this, scheduler will see that the jobs are queued and schedule them again right? Am I missing something important?

File: src/server/reply_send.c [hunk 1]
```diff
@@ -279,6 +279,10 @@ reply_send(struct batch_request *request)
 	int		    rc = 0;
 	int		    sfds = request->rq_conn;		/* socket */
 
+	/* No need to send reply for asynrunjob */
+	if (request->rq_type == PBS_BATCH_AsyrunJob)
+		return 0;
```
Comment (hirenvadalia) [inline src/server/reply_send.c]: we need free request before returning

File: src/server/reply_send.c [hunk 2]
```diff
@@ -362,6 +366,10 @@ reply_ack(struct batch_request *preq)
 	if (preq == NULL)
 		return;
 
+	/* No need to send ack for asynrunjob */
+	if (preq->rq_type == PBS_BATCH_AsyrunJob)
+		return;
```
Comment (hirenvadalia) [inline src/server/reply_send.c]: same as above

File: src/server/reply_send.c [hunk 3]
```diff
@@ -444,6 +452,10 @@ req_reject(int code, int aux, struct batch_request *preq)
 	if (preq == NULL)
 		return;
 
+	/* No need to send reject for asynrunjob */
+	if (preq->rq_type == PBS_BATCH_AsyrunJob)
+		return;
```
Comment (hirenvadalia) [inline src/server/reply_send.c]: same as above

**Comment by bhroam** [comment]
@agrawalravi90 Arun is correct.  The runjob hook needs to return back to the scheduler.  The runjob hook was designed to have the scheduler to not use those resources.  It also returns back a message that gets added to the comment/log.  This reply has to get back to the scheduler.

**Comment by arungrover** [comment]
> 
> Thanks for your reply Arun. If I understand your concern correctly, then this is similar to the mom rejecting a job, which will cause the same issue today, so is that a big issue? The next sched cycle, which will hopefully be sooner after this, scheduler will see that the jobs are queued and schedule them again right? Am I missing something important?

I think it is a little different than a job getting rejected by a mom hook. The reason being, if a mom hook rejects a job, the job is penalized by increasing the run count and eventually such jobs are held by the system. But if a runjob hook rejects the job and this information is not passed on to the scheduler then in every cycle scheduler will take the wrong decision in every cycle by assuming that the job ran and will continue to do so because such a job will not be held automatically.

**Comment by agrawalravi90** [comment]
Thanks for the inputs guys, I understand now that I'll need to come up with a workaround for the runjob hook issue. As you guys suggested, maybe we can make this configurable, tell scheduler to care/not care about runjob hooks and admins can decide if they want to speed things up by telling sched to not care. I would also like to see if we can hold the rejected job for 1/few cycle(s) so that other jobs can use those resources, see if it affects utilization in, and if it is ok to do this. Maybe there are some other ways to get around this as well.

So I'll mark this [WIP] for now.

**Comment by agrawalravi90** [comment]
Started a forum discussion here: http://community.pbspro.org/t/scheduler-spends-94-of-its-time-waiting-for-job-run-ack/2053

**Comment by agrawalravi90** [comment]
The discussion is going in a slightly different direction, a new sched attribute also needs to be created, so I'll close this PR and raise a new one when we've finalized the design

**Pull request closed without merging**